### DISCUSSION TITLE
What happens if the json string is not padded?

### DISCUSSION THREAD

**Discussion by Gun9niR** [simdjson/simdjson#2195] [Q&A] [answered]
In my use case it's difficult to pad the json without allocating a new buffer and copying to it, but my application is performance-critical so the allocation overhead is significant. So the questions are: 

1. Is there any way to work with unpadded string without allocation?
2. And if there's not, what happens if I call `parser::iterate` on an unpadded string? Because surprisingly I've been playing with the parser without padding and there seems to be no error.

**Comment by jkeiser** [comment, marked as answer]
Do you control the initial allocation of the buffer you read your JSON into, and can you pad that initial allocation a bit to match?

There isn't currently a great workaround -- you can use the fallback parser but it doesn't use simd and is slower. There are a couple of reasons we need padding, but one important reason is that we can do faster string processing by reading 8-64 bytes at a time, including when we are at the end of the string (we just retroactively cap the string when we find the end quote).

**Reply by Gun9niR** [reply]
> Do you control the initial allocation of the buffer

That's the problem - I don't.

> use the fallback parser

Is it one of the other open-source libraries such as yyjson?

**Reply by jkeiser** [reply]
No, it was written as part of simdjson. We have multiple backends / kernels for different architectures--arm, Intel with 128-, 256- and 512-bit SIMD, powerpc, and generic 64-bit fallback that runs on anything. Generally we pick the fastest one for your machine, but you can override it. Your application code using simdjson doesn't change, you just switch out kernels.

**Reply by 0AyanamiRei** [reply]
Why does adding free padding allow reading 8 to 64 bytes at once? I remember that simdjson's free padding is a fixed word length. Is this a conservative design?

**Comment by lemire** [comment]
> And if there's not, what happens if I call parser::iterate on an unpadded string? Because surprising I've been playing with the parser without padding and there seems to be no error.

It is safe as long as the end of the string is at least `simdjson::SIMDJSON_PADDING` bytes away from the end of a page (pages are 4kB under Windows, but larger on many other systems). I will actually add a remark to our documentation just now.

**Reply by Gun9niR** [reply]
Thank you for elaborating!

> I've been playing with the parser without padding and there seems to be no error.

I didn't do any special handling for page boundary so I guess it's just luck :p

**Reply by jkeiser** [reply]
Yeah, segfault from accessing that SIMDJSON_PADDING bytes is the only potentially bad outcome I know of. You can even store other stuff in the padding bytes if you want. simdjson does not write to the padding or care about its contents, but it *does* read the bytes from memory as part of its main loop. (This allows us to read the document N bytes at a time all the way to the end.)

**Comment by Gun9niR** [comment]
Thank you both for the explanation!

I have another question that could use your help: I can mitigate my problem by switching to a faster allocator for the padded string (but other allocations in the parser are still std `malloc`). However, when I pass the padded buffer size to the `iterate` function, I still see a lot of `malloc` calls, but they go away if I pass the **unpadded** buffer size.

So the code without those `malloc`s is here, where `m_compressedLength` is just the length of the json string.

```cpp
auto doc = parser.iterate(m_compressedPadded, m_compressedLength, m_compressedLength);
```

The code with `malloc` is

```cpp
auto doc = parser.iterate(m_compressedPadded, m_compressedLength, m_compressedLength + SIMDJSON_PADDING);
```
 
What am I doing wrong here?

**Reply by jkeiser** [reply]
A simdjson parser only allocates if the document being encountered is bigger than any it's encountered before. Basically it grows its buffers but never shrinks them, so that you reach a steady state. If you're doing the second line before the first one, that's what will happen.

If you're not sure, you can check this by telling simdjson not to allocate by preallocating and preventing growth:

```c++
parser.allocate(m_compressedLength); // Preallocate to max capacity
parser.set_max_capacity(m_compressedLength); // Prevent growth
```

After that, do your parser.iterate() calls and they will either run without allocating, or return an error (if the document is bigger than max capacity). That should at least tell you that document size increase is indeed what's happening.

**Comment by lemire** [comment]
@jkeiser @Gun9niR
 
Please see...
https://github.com/simdjson/simdjson/pull/2196

I provide a complete example.

**Reply by Gun9niR** [reply]
That's a neat trick!

**Reply by lemire** [reply]
@Gun9niR I wish we could use it generally but there is much resistance to it. Yet, as a user of the library, you can definitively do it.
